drivers/kscan/kscan_cst816s.c

```diff
@@ -0,0 +1,317 @@
+
+		k_msleep(CST816S_RESET_DELAY);
+		gpio_pin_set_dt(&cfg->rst_gpio, 0);
+	}
+}
+
+static int cst816s_chip_init(const struct device *dev)
+{
+	const struct cst816s_config *cfg = dev->config;
+	int ret = 0;
+	uint8_t chip_id = 0;
+
+	cst816s_chip_reset(dev);
+
+	if (!device_is_ready(cfg->i2c.bus)) {
+		LOG_ERR("I2C bus %s not ready", cfg->i2c.bus->name);
+		return -ENODEV;
+	}
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, CST816S_REG_CHIP_ID, &chip_id);
+	if (ret < 0) {
+		LOG_ERR("failed reading chip id");
+		return ret;
+	}
+
+	if (chip_id != CST816S_CHIP_ID) {
+		LOG_ERR("CST816S wrong chip id: returned 0x%x", chip_id);
+		return -ENODEV;
+	}
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c,
+				     CST816S_REG_IRQ_CTL,
+				     CST816S_IRQ_EN_TOUCH | CST816S_IRQ_EN_CHANGE,
+				     CST816S_IRQ_EN_TOUCH | CST816S_IRQ_EN_CHANGE);
+	if (ret < 0) {
+		LOG_ERR("Could not enable irq");
+		return ret;
+	}
+	return ret;
+}
+
+static int cst816s_init(const struct device *dev)
+{
+	const struct cst816s_config *config = dev->config;
+	struct cst816s_data *data = dev->data;
+
+	data->dev = dev;
+	k_work_init(&data->work, cst816s_work_handler);
+
+#ifdef CONFIG_KSCAN_CST816S_INTERRUPT
+	int ret;
+
+	if (!device_is_ready(config->int_gpio.port)) {
+		LOG_ERR("GPIO port %s not ready", config->int_gpio.port->name);
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+	if (ret < 0) {
+		LOG_ERR("Could not configure interrupt GPIO pin");
+		return r;
+	}
+
+	ret = gpio_pin_interrupt_configure_dt(&config->int_gpio,
+			GPIO_INT_EDGE_TO_ACTIVE);
+	if (ret < 0) {
+		LOG_ERR("Could not configure interrupt GPIO interrupt.");
+		return r;
+	}
+
+	gpio_init_callback(&data->int_gpio_cb, cst816s_isr_handler,
+			   BIT(config->int_gpio.pin));
+#else
+	k_timer_init(&data->timer, cst816s_timer_handler, NULL);
+#endif
+
+	return cst816s_chip_init(dev);
+}
+
+static const struct kscan_driver_api cst816s_driver_api = {
+	.config = cst816s_configure,
+	.enable_callback = cst816s_enable_callback,
+	.disable_callback = cst816s_disable_callback,
+};
+
+#define CST816S_DEFINE(index)                                                 \
+	static const struct cst816s_config cst816s_config_##index = {         \
+		.i2c = I2C_DT_SPEC_INST_GET(index),                           \
+		COND_CODE_1(CONFIG_KSCAN_CST816S_INTERRUPT,                   \
+			   (.int_gpio = GPIO_DT_SPEC_INST_GET(i, irq_gpios),),\
+			   ())                                                \
+		.rst_gpio = COND_CODE_1(DT_INST_NODE_HAS_PROP(index, rst_gpios),\
+				(GPIO_DT_SPEC_INST_GET(index, rst_gpios)), ()) \
+	}                                                                    \
+	static struct cst816s_data cst816s_data_##index;                     \
+	DEVICE_DT_INST_DEFINE(index, cst816s_init, NULL,                     \
+			     &cst816s_data_##index, &cst816s_config_##index,          \
+			     POST_KERNEL, CONFIG_KSCAN_INIT_PRIORITY,                 \
+			     &cst816s_driver_api);
```

Comment:
suggestion: format should be improved here, you may use https://github.com/zephyrproject-rtos/zephyr/pull/41307